libbeat/mime/byte.go

```diff
@@ -0,0 +1,71 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package mime
+
+import (
+	"encoding/json"
+	"encoding/xml"
+	"net/http"
+	"strings"
+
+	"github.com/h2non/filetype"
+)
+
+const (
+	// size for mime detection, office file
+	// detection requires ~8kb to detect properly
+	headerSize = 8192
+)
+
+// DetectBytes tries to detect a mime-type based off
+// of a chunk of bytes passed into the function
+func DetectBytes(data []byte) string {
+	header := data
+	if len(data) > headerSize {
+		header = data[:headerSize]
+	}
+	kind, err := filetype.Match(header)
+	if err == nil && kind != filetype.Unknown {
+		// we have a known filetype, return
+		return kind.MIME.Value
+	}
+	// if the above fails, try and sniff with http sniffing
+	netType := http.DetectContentType(header)
+	// try and parse any sort of text as json or xml
+	if strings.HasPrefix(netType, "text/plain") {
+		if detected := detectEncodedText(data); detected != "" {
+			return detected
+		}
+	}
+	if netType == "application/octet-stream" {
+		return ""
```

Comment:
Are we not reporting "application/octet-stream" because we didn't decode it and look?  I'm wondering if it is still  useful to report what we do know.

Reply:
so `application/octet-stream` basically is the same as saying something is "bytes" -- it's a generic fallback mimetype -- so the thought was that rather than using it, we'd just say we couldn't determine the type.

Reply:
btw, here's where `octet-stream` actually gets returned https://golang.org/src/net/http/sniff.go#L37 -- it's basically set if the http sniffer can't match any known types.

Reply:
ah ok.  Do you mind adding a comment calling out net/nttp/sniff.go behavior?  I'm just thinking if we get a bug report about us not reporting "application/octet-stream" a year from now it will help us remember why we are doing this.